### PR TITLE
Add mini-language standalone parser for observe 

### PR DESCRIPTION
**Pull request by kitchoi**
Part of item 5 in #977

This PR:
- Adds the grammar file for generating a standalone parser using `lark-parser`
- Adds the generated parser
- Add the developer tool for regenerating the parser whenever the grammar is modified. This introduces a (build) dependency on `lark-parser`.

Tests are added to verify the parser can differentiate good texts versus bad ones.  The parser converts text into a syntax tree.  Further interpretation on the syntax tree will require item 4 (`Expression`) in #977, which has not been introduced yet. But here is an example of how it will get used:
https://github.com/enthought/traits/blob/318f28485211f30ab18311a05ef47d4c63718336/traits/observers/parsing.py#L139-L160

The standalone parser is only for internal use. There will be a user-facing one later when `Expression` is finalized.

**Checklist**
- [x] Tests
- ~Update API reference (`docs/source/traits_api_reference`)~
- ~Update User manual (`docs/source/traits_user_manual`)~
- ~Update type annotation hints in `traits-stubs`~

### CODE REVIEW

**Comment by kitchoi** [comment]
The generated 2000+ lines parser file `_generated_parser.py` is probably NOT reviewable.

File: etstool.py [hunk 1]
```diff
@@ -385,6 +386,38 @@ def test_all(edm):
...
+    parameters["grammar_path"] = os.path.join(
+        observers_dir, "_dsl_grammar.lark")
+
+    command = (
+        "{edm} run -e {environment} -- "
+        "python -m lark.tools.standalone "
+        "{grammar_path}"
+    )
+
+    with open(out_path, "w") as out_file:
```
Comment (corranwebster) [inline etstool.py:416]: Is there a command-line option to get `lark` to output to a file?  that would be preferable to capturing stdout.

Comment (kitchoi) [reply etstool.py:416]: I checked and unfortunately no:
https://github.com/lark-parser/lark/blob/master/lark/tools/standalone.py#L87

Well spotted about the source distribution.  I will add the file to MANIFEST.in

File: traits/observers/_dsl_grammar.lark [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// Grammar for Traits Mini Language used in observe
```
Comment (kitchoi) [inline traits/observers/_dsl_grammar.lark:1]: Just realized nowhere in the file says what this lark file is for (it is kind of implicitly there in the filename/context). So I added this line.

Comment (mdickinson) [reply traits/observers/_dsl_grammar.lark:1]: Could we also record somewhere the exact command-line command needed to generate the parser from the grammar? (We probably eventually want that wrapped up in an `etstool` task, too, but that's separate.)

Comment (mdickinson) [reply traits/observers/_dsl_grammar.lark:1]: Sorry, ignore. I should have looked harder at the PR first. (/me crawls back into his hole)

File: etstool.py (hunk 1 above)
Comment (mdickinson) [inline etstool.py:416]: Please include an explicit encoding here.

Comment (mdickinson) [reply etstool.py:416]: (related: https://www.python.org/dev/peps/pep-0597/)

**Comment by mdickinson** [comment]
It appears that the lark output can differ from run to run: if I do `python etstool.py generate-parser`, I get a big diff in the `DATA` and `MEMO` constants from the existing `_generated_parser.py`.

Do you know if there's any way to control this? Reproducibility would be good.

**Comment by mdickinson** [comment]
> Do you know if there's any way to control this?

From https://github.com/lark-parser/lark/issues/278, it looks as though the answer for now is "not really, no".

File: etstool.py (hunk 1 above)
Comment (kitchoi) [reply etstool.py:416]: Well spotted, thanks.

File: traits/observers/_dsl_grammar.lark (hunk 2 above)
Comment (kitchoi) [reply traits/observers/_dsl_grammar.lark:1]: Not at all, you helped me realize that I misspelt the filename for `etstool.py` as `etstools.py` on line 3. Now fixed.  Waiting for CI...

**Comment by mdickinson** [comment]
> From [lark-parser/lark#278](https://github.com/lark-parser/lark/issues/278), it looks as though the answer for now is "not really, no".

Hmm. But that issue's closed, and apparently fixed. I'm not sure what's going on here.

**Comment by kitchoi** [comment]
> It appears that the lark output can differ from run to run: if I do python etstool.py generate-parser, I get a big diff in the DATA and MEMO constants from the existing _generated_parser.py.

I can reproduce the different outputs with lark 0.8, Python 3.8 (where dictionaries are ordered), even with PYTHONHASHSEED set to a fixed value. No idea why.

**Comment by kitchoi** [comment]
Merging. It is not clear if there is a trivial way to get the output consistent and it does not seem important enough to block the parser from going in...

**Comment by mdickinson** [comment]
> It is not clear if there is a trivial way to get the output consistent and it does not seem important enough to block the parser from going in...

Agreed, but please could you open an issue? We may want to report upstream too, but a local issue would be the first step for that.

**Comment by kitchoi** [comment]
I typed and removed ("will open an issue upstream") because I did not want to promise that - I will need to work out a reproducible example first.

**Comment by mdickinson** [comment]
Yep, just a local issue would be fine; that issue will act as a trigger for us to send an upstream issue when we're ready.

**Comment by kitchoi** [comment]
Opened https://github.com/lark-parser/lark/issues/584

I think https://github.com/lark-parser/lark/issues/278 is about the parsed syntax tree showing nondeterministic behaviour.

Here the nondeterministic behaviour is about the generated code not being static, creating noisy diff despite using the same grammar.

**Comment by mdickinson** [comment]
@kitchoi Thank you!

One option we might consider is not tracking the parser code in version control, but instead generating it at package install time (i.e., as a result of running the `setup` script). I suspect that that would be more annoying than helpful, but there are tradeoffs either way.

**Pull request merged**